### DISCUSSION TITLE
State.Empty not recomposing

### DISCUSSION THREAD

**Discussion by georgeBanno** [coil-kt/coil#853] [Q&A]
I've been working with the Coil Composable today and have seen the following behavior.  I'm sure there is a good reason for this that I'm missing. However,  I've reviewed the code and can't really see why this is happening.

Whenever painter.state is compared to ImagePainter.State.Empty, like below, the composable never recomposes to show the image or the progress dialog.  However if I remove the Empty state check the progress indicator displays and then the image is displayed.  I'd love to know why this is. 

This code will always show the text  "Oops I'm Empty" .
```
@ExperimentalCoilApi
@Composable
fun ShowAPIImage(url: String, imageLoader: ImageLoader) {
    val painter = rememberImagePainter(data = url, imageLoader = imageLoader)

    Log.e("Composing", "I'm composing ${painter.state}")
    when (painter.state) {
        ImagePainter.State.Empty -> {
            Text(text = "Oops I'm Empty")
        }
        is ImagePainter.State.Error -> {
            Text(text = "Oops I'm broke")
        }
        is ImagePainter.State.Loading -> {
            CircularProgressIndicator()
        }
        else -> {
            Image(
                painter = painter,
                contentDescription = "Image",
            )
        }
    }
}

```

This code will always first show the progress indicator and then the image when it loads.
```
@ExperimentalCoilApi
@Composable
fun ShowAPIImage(url: String, imageLoader: ImageLoader) {
    val painter = rememberImagePainter(data = url, imageLoader = imageLoader)

    Log.e("Composing", "I'm composing ${painter.state}")
    when (painter.state) {
        is ImagePainter.State.Error -> {
            Text(text = "Oops I'm broke")
        }
        is ImagePainter.State.Loading -> {
            CircularProgressIndicator()
        }
        else -> {
            Image(
                painter = painter,
                contentDescription = "Image",
            )
        }
    }
}

```

**Comment by colinrtwhite** [comment]
`rememberImagePainter` won't launch the image request until it has a target size and it won't have a size until the `Image` composable is in the UI tree. In the first example, the `Image` is never added to the tree. In the second example, the `Image` is added to the tree and the image request is started. If you want you can manually set the image request's size using `ImageRequest.Builder.size(OriginalSize)`. This is due to [a compose design decision](https://issuetracker.google.com/186012457).

**Reply by georgeBanno** [reply]
Thanks for the reply. I tried setting the size on the modifier for the image but it didn't seem to make a difference when I'm checking for Empty state. Even with the size set via the modifier the only way the image will show is by removing the code below from the when statement.  Looking at the logs it never makes a request to the URL. 

Code for setting size 
````
 Image(
                painter = painter,
                contentDescription = "Image",
                modifier = Modifier.width(200.dp).height(100.dp)
            )
````

Remove this code and it works with or without the size provided:
`ImagePainter.State.Empty -> {
            Text(text = "Oops I'm Empty")
        }`
